### PR TITLE
feat: added new extra for get size of objects when collect metrics

### PR DESCRIPTION
**Pull request by tofuurem**

### CODE REVIEW

File: blocks/types/metrics.py [hunk 1]
```diff
@@ -16,21 +17,45 @@ def interval(self) -> float:
...
     max_processing: float
     min_processing: float
     avg_processing: float
+    max_size: float
+    min_size: float
+    avg_size: float
 
     def __str__(self) -> str:
-        return '[{0} - {1}] Processed {2} events for {3}. MIN: {4}, MAX: {5}, AVG: {6}'.format(
+        return '[{0} - {1}] Processed {2} events for {3}.({4}, {5}, {6}) ({7}, {8}, {9})'.format(
```
Comment (AntonNemtsev) [inline blocks/types/metrics.py:50]: Maybe point out more explanations?

Comment (tofuurem) [reply blocks/types/metrics.py:50]: This is generally a basic line, if logs are needed I find it easier to customize in the processor )

Comment (AntonNemtsev) [reply blocks/types/metrics.py:50]: ok

**Review by AntonNemtsev — APPROVED** [review]

File: blocks/metric_collector.py [hunk 2]
```diff
@@ -49,15 +49,22 @@ def __init__(self, metric_time_interval: int) -> None:
...
         """
 
         self._tc = metric_time_interval
-        self._collected_data: Dict[Processor, Dict[Event, List[EventTime]]] = OrderedDict()
+        self._collected_data: Dict[Processor, Dict[Event, List[MetricEvent]]] = OrderedDict()
 
         self.timer = _ProcTimer()
 
-    def collect(self, processor_type: Processor, event_time: EventTime) -> None:
+    def collect(self, processor_type: Processor, event_time: MetricEvent) -> None:
```
Comment (tribunsky-kir) [inline blocks/metric_collector.py:67]: I suggest to rename `event_time` arguments. As I see in `runners.py` it is not just telemetry/tracing info by design, even if size feature is unavailable.

Wouldn't it better to split memory tracker and duration meter?

File: blocks/runners.py [hunk 3]
```diff
@@ -12,10 +12,17 @@
...
 from blocks.types.base import is_named_tuple
-from blocks.metric_collector import MetricCollector, EventTime
+from blocks.metric_collector import MetricCollector, MetricEvent
 
 SyncProcessors = DefaultDict[Type[Event], List[Processor]]
 
+try:
+    import objsize
+
+    METRIC_SIZE: bool = True
```
Comment (tribunsky-kir) [inline blocks/runners.py:22]: _stylish_: set `METRIC_SIZE` to `True` in try/except/**else**.

File: blocks/runners.py [hunk 4]
```diff
@@ -150,7 +161,15 @@ def _handle_pool_exceptions(self, exc: BaseException, parallel_event: ParallelEv
...
         with self._mp.timer as timer:
             output_event = processor(input_event)
-        self._mp.collect(processor, EventTime(type(input_event), timer.start, timer.end))
+        self._mp.collect(
+            processor,
+            MetricEvent(
+                type(input_event),
+                timer.start,
+                timer.end,
+                size_object=round(objsize.get_deep_size(processor) / (1<<20), 3) if METRIC_SIZE else 0
```
Comment (tribunsky-kir) [inline blocks/runners.py:170]: Not great deal, but looks like it is possible to not degrade performance on every call if `objsize` isn't available.

E.g. move size calculation to separate method which is resolved only once.

File: blocks/metric_collector.py (hunk 2 above)
Comment (tribunsky-kir) [reply blocks/metric_collector.py:67]: **UPDATE:** _ProcTimer is not great name also for keeping both timing and memory.

**Pull request closed without merging**